### PR TITLE
ZIO channel and attribute rework WIP

### PR DESCRIPTION
**Pull request by pollend**
I've been looking at the zio api and I have a rework of the api that defines all the attributes and channels through anonymous nested structs. you use ZIO_ATTR(id) and ZIO_CHANNEL(id) to retrieve the accessor type. attributes can either be nested within the definition of a channel or at the root of the api struct for device attribute configurations.  At the moment this does not quite work and I was looking for some suggestions. I realize this is a pretty large departure from the original api.

I think this could address some of my problems with the current zio sensor api. The definitions and  for attributes and channels are more defined. this removes the need for the setter functions in the base api. I was also looking into storing the attributes and channels into nested arrays but I think this is more defined. I'm not sure about using the preprocessor in this way, but I think these could be tweaked and reworked.

The majority of the interesting changes are in synth.c and synth.h file.

### CODE REVIEW

**Review by dbkinder — COMMENTED** [review]
I don't see any docs (.rst file) for review, I suspect because zio isn't a public interface. Is there a documentation impact for this change?

**Comment by pollend** [comment]
This is just an early wip. Its a significant change from where zio was going at the moment. I kind want to see what people think about this change first before putting in significant effort to develop something more complete.

**Comment by dbkinder** [comment]
OK.  I also see a PR for Documentation for zio #16384

**Comment by microbuilder** [comment]
@dbkinder Yeah, I made a boilerplate PR just to get started, but haven't invested time into documenting the details until the dust settles on this a bit.

File: include/zio/zio_dev.h [hunk 1]
```diff
@@ -129,75 +177,76 @@ struct zio_dev_chan {
...
-typedef int (*zio_dev_get_attr_t)(struct device *dev, uint32_t attr_idx,
-				  struct zio_variant *attr);
+typedef int (*zio_get_attr_t)(struct device *dev, const u16_t channel_idx, const u16_t attribute_idx, struct zio_variant *attr);
+
+
+/** Attribute record. */
+struct zio_attr_desc {
+	/* Attribute data type describing the expected data type. */
+	enum zio_variant_type data_type;
+	zio_set_attr_t set_attr;
```
Comment (bfrog) [inline include/zio/zio_dev.h]: These two function pointers are I think the most significant part of this change set and I definitely agree that they make the device driver API clearer, avoids unnecessary switch/case function wrappers, and generally makes for a nicer API

File: include/zio/zio_dev.h [hunk 2]
```diff
@@ -260,19 +309,27 @@ typedef int (*zio_dev_attach_buf_t)(struct device *dev, struct zio_buf *buf);
  */
 typedef int (*zio_dev_detach_buf_t)(struct device *dev);
 
+/**
+ * @brief Function to set a device attribute
+ *
+ * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ */
+typedef int (*zio_dev_set_attr_t)(struct device *dev, const u16_t attr_idx, const struct zio_variant var);
```
Comment (bfrog) [inline include/zio/zio_dev.h]: This seems like it'd be unnecessary given each attribute description has a get/set function pointer

File: include/zio/zio_dev.h [hunk 3]
```diff
@@ -129,75 +177,76 @@ struct zio_dev_chan {
...
- */
-typedef int (*zio_chan_set_attr_t)(struct device *dev, const u32_t chan_idx,
-				   const u32_t attr_idx,
-				   const struct zio_variant var);
 /**
  * @brief Function to get a channel attribute for a device
  *
  * @return -EINVAL if an invalid attribute id was given, 0 otherwise
  */
 typedef int (*zio_chan_get_attr_t)(struct device *dev, const u32_t chan_idx,
```
Comment (bfrog) [inline include/zio/zio_dev.h]: I don't think this, or any of the other get/set attr function pointer defs are needed anymore with your change, where each attribute has its own getter/setter

**Review by teburd — COMMENTED** [review]
I think the general idea of putting the setter/getter function pointers into the attribute description makes a lot of sense. What to pass into those function pointers is another issue, the callback might just need a parameter to give it the context it needs. I don't think CONTAINER_OF would solve the problem, you would need something like it that could understand pointer math for array members where the ptr is pointing to one thing in that array. Maybe that already exists somewhere.

Another thing I would keep in mind is that the current API has some faults I think, we don't support the idea of related sets of channels where for example the x,y,z accelerometer channels of a 9 axis sensor like the one I believe your writing a driver for share a lot of attributes such as value ranges (+/- 2g, 8g, 16g, etc).

This changeset seems like its on the right track to creating the building block to help solve that problem as well.

**Comment by pfalcon** [comment]
@bfrog: Some feedback re: https://github.com/zephyrproject-rtos/zephyr/commit/4fe881977bf78a5acba1ba6db1e4e5709e8b4cd9 : 

> Add a new sensor device API called zio_dev

I find "zio" naming to be confusing. Which letter exactly in "zio" stands for "sensor"? And why someone would need to decipher letters, why not call it with "sensor" in name for clarity? We already have one cryptic, confusing, conflicting name like that: https://github.com/zephyrproject-rtos/zephyr/pull/16114 .

**Comment by bfrog** [comment]
> @bfrog: Some feedback re: [4fe8819](https://github.com/zephyrproject-rtos/zephyr/commit/4fe881977bf78a5acba1ba6db1e4e5709e8b4cd9) :
> 
> > Add a new sensor device API called zio_dev
> 
> I find "zio" naming to be confusing. Which letter exactly in "zio" stands for "sensor"? And why someone would need to decipher letters, why not call it with "sensor" in name for clarity? We already have one cryptic, confusing, conflicting name like that: #16114 .

I don't disagree, I needed a name, and I wanted something short, zephyr io abbreviated to zio seemed easy enough.

The general problem with naming it sensors is that this API is meant to encompass more of what IIO does in Linux, provide an input/output stream with attributes. I'm open to suggestions.

**Comment by mbolivar** [comment]
> The general problem with naming it sensors is that this API is meant to encompass more of what IIO does in Linux, provide an input/output stream with attributes and triggers. I'm open to suggestions.

I'm still +1 on rtio for real/time input/output.

I believe that this I/O framework is meant to cover actuators as well, so I disagree that "sensor" should be in the name, as that would be misleading. If I'm wrong about that, then never mind.

**Comment by WilliamGFish** [comment]
Hi All,
I have come across this whilst working on a implementation of the BT Mesh Sensor Model which has a requirement for exposing sensor descriptions and attributes which appears to have a significant overlap.

I'm not sure if anyone has reviewed the BT documentation to consider potential relevance? I think it would valuable to have a look.

Billy..

**Comment by mbolivar** [comment]
The initial ZIO work has since been replaced by RTIO; see the branch by the same name on @bfrog's zephyr fork.

**Comment by WilliamGFish** [comment]
> The initial ZIO work has since been replaced by RTIO; see the branch by the same name on @bfrog's zephyr fork.

Thanks, I'll have a look.

**Comment by bfrog** [comment]
I think since zio has become a dead end, it might be worth considering closing this.

**Pull request closed without merging**